### PR TITLE
[iov-ledger-bns] Track Ledger state

### LINKED ISSUE #218: Implement LedgerSimpleAddressKeyringEntry.canSign
as discussed in https://github.com/iov-one/web4/pull/214

### PR DESCRIPTION
**Pull request by webmaster128**
Resolves #218

### CODE REVIEW

File: custom_types/ledgerhq__hw-transport-node-hid/index.d.ts [hunk 1]
```diff
@@ -1 +1,38 @@
-declare module "@ledgerhq/hw-transport-node-hid";
+declare module "@ledgerhq/hw-transport-node-hid" {
+  import { Device, HID } from "node-hid";
+
+  /**
+   * "A Descriptor is a parametric type that is up to be determined for the
+   * implementation. it can be for instance an ID, an file path, a URL,..."
```
Comment (rudi-cilibrasi) [inline custom_types/ledgerhq__hw-transport-node-hid/index.d.ts]: `a file path`

Comment (webmaster128) [reply custom_types/ledgerhq__hw-transport-node-hid/index.d.ts]: haha, nice catch. You need to add an upstream PR to fix the source of that quote ;)

Comment (webmaster128) [reply custom_types/ledgerhq__hw-transport-node-hid/index.d.ts]: fixed

File: custom_types/ledgerhq__hw-transport-node-hid/index.d.ts [hunk 2]
```diff
@@ -1 +1,38 @@
-declare module "@ledgerhq/hw-transport-node-hid";
+declare module "@ledgerhq/hw-transport-node-hid" {
+  import { Device, HID } from "node-hid";
+
+  /**
+   * "A Descriptor is a parametric type that is up to be determined for the
+   * implementation. it can be for instance an ID, a file path, a URL,..."
```
Comment (rudi-cilibrasi) [inline custom_types/ledgerhq__hw-transport-node-hid/index.d.ts]: `. It`

File: packages/iov-ledger-bns/package.json [hunk 3]
```diff
@@ -22,8 +22,8 @@
     "prebuild": "yarn format && yarn lint",
     "build": "tsc && rm ./types/* && mv build/*.d.ts ./types",
     "build-or-skip": "[ -n \"$SKIP_BUILD\" ] || yarn build",
-    "listen": "node ./build/scripts/listen.js | grep -v '^dev'",
-    "checkApp": "node ./build/scripts/checkapp.js | grep -v '^dev'"
+    "listen": "ts-node ./src/scripts/listen.ts | grep -v '^dev'",
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-ledger-bns/package.json]: Yes, here: https://github.com/LedgerHQ/ledgerjs/issues/204#issuecomment-413774755 Issue on node-hid is not yet written

File: packages/iov-ledger-bns/src/statetracker.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import { connectToFirstLedger } from "./exchange";
+
+export enum LedgerState {
+  Disconnected,
+  Connected,
+  IovAppOpen,
+}
+
+export class StateTracker {
+  private static async checkConectedAndAppOpen(): Promise<LedgerState> {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-ledger-bns/src/statetracker.ts]: I wonder what a Transport really is. Given that I sometimes got the response of one transport in another transport, I don't think there is a lot of overhead in creating one. But yeah, maybe we could pass around one Transport.

File: packages/iov-ledger-bns/src/app.spec.ts [hunk 5]
```diff
@@ -98,7 +99,7 @@ describe("Sign with ledger app", () => {
   it("is compatible with our codecs", async () => {
     pendingWithoutInteractiveLedger();
 
-    const pubkey = await getPublicKey(transport);
+    const pubkey = await getPublicKey(transport!);
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-ledger-bns/src/app.spec.ts]: :)

File: packages/iov-keycontrol/src/valueandupdates.ts [hunk 6]
```diff
@@ -18,6 +18,23 @@ export class ValueAndUpdates<T> {
     this.producer = producer;
     this.updates = MemoryStream.createWithMemory(this.producer);
   }
+
+  public waitFor(value: T): Promise<void> {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-keycontrol/src/valueandupdates.ts]: After getting started with xstreams, I love the concept of separate streams and producers. In many places we utilize this to make producers private. This way, you always know where the data comes from even of anyone can listen. I did not yet see something similar in rxjs. Exposing the BahaviourSubject means exposing write access to the stream.

[the remark this answers is not in the document]
Comment (ethanfrey) [reply packages/iov-keycontrol/src/valueandupdates.ts]: Cool, then we keep on this path. Thanks for the feedback

**Pull request merged**